SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -0,0 +1,1833 @@
+#include "Utils/ComponentReader.h"
+#include "Utils/RepLayoutUtils.h"
+
+DEFINE_LOG_CATEGORY(LogActorSystem);
+
+namespace SpatialGDK
+{
+struct ActorSystem::RepStateUpdateHelper
+{
+	RepStateUpdateHelper(USpatialActorChannel& Channel, UObject& TargetObject)
+		: ObjectPtr(MakeWeakObjectPtr(&TargetObject))
+		, ObjectRepState(Channel.ObjectReferenceMap.Find(ObjectPtr))
+	{
+	}
+
+	~RepStateUpdateHelper() { check(bUpdatePerfomed); }
+
+	FObjectReferencesMap& GetRefMap()
+	{
+		if (ObjectRepState)
+		{
+			return ObjectRepState->ReferenceMap;
+		}
+		return TempRefMap;
+	}
+
+	void Update(ActorSystem& Actors, USpatialActorChannel& Channel, const bool bReferencesChanged)
+	{
+		check(!bUpdatePerfomed);
+
+		if (bReferencesChanged)
+		{
+			if (ObjectRepState == nullptr && TempRefMap.Num() > 0)
+			{
+				ObjectRepState =
+					&Channel.ObjectReferenceMap.Add(ObjectPtr, FSpatialObjectRepState(FChannelObjectPair(&Channel, ObjectPtr)));
+				ObjectRepState->ReferenceMap = MoveTemp(TempRefMap);
+			}
+
+			if (ObjectRepState)
+			{
+				ObjectRepState->UpdateRefToRepStateMap(Actors.ObjectRefToRepStateMap);
+
+				if (ObjectRepState->ReferencedObj.Num() == 0)
+				{
+					Channel.ObjectReferenceMap.Remove(ObjectPtr);
+				}
+			}
+		}
+#if DO_CHECK
+		bUpdatePerfomed = true;
+#endif
+	}
+
+private:
+	FObjectReferencesMap TempRefMap;
+	TWeakObjectPtr<UObject> ObjectPtr;
+	FSpatialObjectRepState* ObjectRepState;
+#if DO_CHECK
+	bool bUpdatePerfomed = false;
+#endif
+};
+
+struct FSubViewDelta;
+
+ActorSystem::ActorSystem(const FSubView& InSubView, const FSubView& InTombstoneSubView, USpatialNetDriver* InNetDriver,
+						 FTimerManager* InTimerManager, SpatialEventTracer* InEventTracer)
+	: SubView(&InSubView)
+	, TombstoneSubView(&InTombstoneSubView)
+	, NetDriver(InNetDriver)
+	, TimerManager(InTimerManager)
+	, EventTracer(InEventTracer)
+{
+}
+
+void ActorSystem::Advance()
+{
+	for (const EntityDelta& Delta : TombstoneSubView->GetViewDelta().EntityDeltas)
+	{
+		if (Delta.Type == EntityDelta::ADD)
+		{
+			AActor* EntityActor = TryGetActor(
+				UnrealMetadata(SubView->GetView()[Delta.EntityId]
+								   .Components.FindByPredicate(ComponentIdEquality{ SpatialConstants::UNREAL_METADATA_COMPONENT_ID })
+								   ->GetUnderlying()));
+			if (EntityActor == nullptr)
+			{
+				continue;
+			}
+			UE_LOG(LogSpatialReceiver, Verbose, TEXT("The received actor with entity ID %lld was tombstoned. The actor will be deleted."),
+				   Delta.EntityId);
+			// We must first Resolve the EntityId to the Actor in order for RemoveActor to succeed.
+			NetDriver->PackageMap->ResolveEntityActor(EntityActor, Delta.EntityId);
+			RemoveActor(Delta.EntityId);
+		}
+	}
+
+	for (const EntityDelta& Delta : SubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const AuthorityChange& Change : Delta.AuthorityLostTemporarily)
+			{
+				AuthorityLost(Delta.EntityId, Change.ComponentSetId);
```

Comment:
Shouldn't this be followed by an AuthGained() ?

Reply:
Oh I see it
I think i'm gonna be sick
Could we add a log why this is necessary?

Reply:
ha
sure yeah good idea, added a comment